src/Lykke.Job.CandleHistoryWriter.Repositories/HistoryMigration/HistoryProviders/TradesSQLHistory/TradesSqlHistoryRepository.cs

```diff
@@ -52,25 +70,59 @@ string assetPairId
                 
                 using (var sqlCommand = new SqlCommand(BuildCurrentQueryCommand(), _sqlConnection))
                 {
-                    sqlCommand.CommandTimeout = 180; // 3 minutes
+                    sqlCommand.CommandTimeout = (int)_sqlTimeout.TotalSeconds;
                     using (var reader = await sqlCommand.ExecuteReaderAsync())
                     {
                         while (await reader.ReadAsync())
                         {
                             result.Add(new TradeHistoryItem
                             {
                                 Id = reader.GetInt64(0),
-                                Volume = reader.GetDecimal(1),
-                                Price = reader.GetDecimal(2),
-                                DateTime = reader.GetDateTime(3),
-                                OppositeVolume = reader.GetDecimal(4)
+                                AssetToken = reader.GetString(1),
+                                Direction = (TradeDirection)Enum.Parse(typeof(TradeDirection), reader.GetString(2)),
+                                Volume = reader.GetDecimal(3),
+                                Price = reader.GetDecimal(4),
+                                DateTime = reader.GetDateTime(5),
+                                OppositeVolume = reader.GetDecimal(6),
+                                OrderId = Guid.Parse(reader.GetString(7)),
+                                OppositeOrderId = Guid.Parse(reader.GetString(8)),
+                                TradeId = reader.GetString(9)
                             });
                         }
                     }
                 }
 
                 if (result.Count > 0)
                 {
+                    // Now we need to remove the last several trades which have the same date and time (accuracy - to seconds).
+                    // This will guarantee that we did not peek up some orders of the same trade on this iteration, and others
+                    // on the next. On the next iteration we will read them again for the next batch. No temporary buffer, for
+                    // it can't save any observable value of time. NOTE: if we have got less records than _sqlQueryBatchSize,
+                    // this means that we obtained the last )or the single) data pack, and there is no reason to delete "tail"
+                    // trades.
+
+                    if (result.Count == _sqlQueryBatchSize)
+                    {
+                        // I do not want to use Linq.Remove(func) for it might case full sequence iterating which is slower than
```

Comment:
```c#
return result.TakeWhile(c => c.DateTime < lastDateTime).ToArray();
```

instead of all code inside ```if (result.Count == _sqlQueryBatchSize)``` block will be much more clear and easy to maintain. Both versions will spend around millisecond, if not less than, of CPU time, wich is nothing against total migration time. Furthermore, I believe, if return type of the method could be changed to IEnumerable, option that I've suggested will be faster (ToArray() could be removed then). But anyway this speed doesn't matter at all.